### PR TITLE
Motion Detection: Add optical flow subsampler + store numpy arrays in npz metadata

### PR DESCRIPTION
**Pull request by danielmend**
will let us do dense optical flow in the 2nd stage (after downloading)

#### CV2
- [x] implement cv2 optical flow detection
- [x] parameterize the optical flow hyperparams (let user specify)
- [x] figure out file extension when saving optical flow -- ideally we'd want to be able to np.load() the optical flow data directly
- [x] Temporal subsampling
- [x] Test end2end WDS input -> Files
- [x] Test end2end WDS input -> WDS
- [x] Test magnitude metric

### CODE REVIEW

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    def __call__(self, frame1, frame2):
+        return cv2.calcOpticalFlowFarneback(frame1, frame2, None, self.pyr_scale, self.levels, self.winsize, self.iterations, self.poly_n, self.poly_sigma, self.flags)
+        
+class OpticalFlowSubsampler:
+    """
+    Detects optical flow in video frames
+    """
+
+    def __init__(self, detector="cv2", fps=-1, cv2_params=None, raft_params=None):
```
Comment (iejMac) [inline video2dataset/subsamplers/optical_flow_subsampler.py:32]: I think one "params" is fine you won't ever be using both

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                prvs = next_frame
+        except Exception as err:  # pylint: disable=broad-except
+            return [], None, str(err)
+
+        opt_flow = np.array(optical_flow)
+        mean_magnitude_per_frame = np.linalg.norm(opt_flow, axis=-1).mean(axis=(1, 2))
+        mean_magnitude = float(mean_magnitude_per_frame.mean())
+
+        metrics = [mean_magnitude, mean_magnitude_per_frame.tolist()]
+        return opt_flow.astype(np.float16), metrics, None
```
Comment (iejMac) [inline video2dataset/subsamplers/optical_flow_subsampler.py:116]: maybe astype shouldn't be default. Maybe we include this as a downsample option along with resize

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+                flow = self.detector(prvs, next_frame)
+
+                optical_flow.append(flow)
+                prvs = next_frame
+        except Exception as err:  # pylint: disable=broad-except
+            return [], None, str(err)
+
+        opt_flow = np.array(optical_flow)
+        mean_magnitude_per_frame = np.linalg.norm(opt_flow, axis=-1).mean(axis=(1, 2))
```
Comment (iejMac) [inline video2dataset/subsamplers/optical_flow_subsampler.py:112]: make like a "downsampling" section and put that in there. We want to do this every time though and other things only when parameters indicate

File: video2dataset/workers/optical_flow_worker.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    Returns:
+        bytes: A bytestring representing the compressed numpy arrays.
+    """
+
+    stream = io.BytesIO()
+    np.savez_compressed(stream, **numpy_dict)
+    return stream.getvalue()
+
+
+def frames_to_mp4_bytes(frames, fps=30, codec="mp4v"):
```
Comment (iejMac) [inline video2dataset/workers/optical_flow_worker.py:35]: why do we need this?

File: video2dataset/workers/optical_flow_worker.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        with open(video_path, "rb") as f:
+            video_bytes = f.read()
+
+    # Remove the temporary video file
+    os.remove(video_path)
+
+    return video_bytes
+
+
+def convert_frames_depth(frames, target_depth=np.uint8):
```
Comment (iejMac) [inline video2dataset/workers/optical_flow_worker.py:75]: same why do we need this?

File: video2dataset/workers/optical_flow_worker.py (hunk 4 above)
Comment (danielmend) [reply video2dataset/workers/optical_flow_worker.py:35]: i cant leave the array of frames in streams it wont get written to mp4  -- i need to re-encode as bytes before i pass to the sample writer i thought

File: video2dataset/workers/optical_flow_worker.py (hunk 5 above)
Comment (danielmend) [reply video2dataset/workers/optical_flow_worker.py:75]: converts stuff to uint8 so that it can be written as an mp4 bytestream

**Comment by iejMac** [comment]
lgtm, we can iron out the last few things in future PR's

**Pull request merged**